package/MDAnalysis/analysis/waterdynamics.py

```diff
@@ -1197,96 +1199,72 @@ class SurvivalProbability(object):
+    tau_max : int
+      Survival probability is calculated for the range :math:`1 <= \tau <= tau_max`
+    verbose : Boolean
+      If True, prints progress and comments to the console.
+
+    Returns
+    -------
+      tau_timeseries : list
+        tau from 1 to tau_max
+      sp_timeseries : list
+        survival probability for each value of `tau`
 
     .. versionadded:: 0.11.0
 
     """
 
-    def __init__(self, universe, selection, t0, tf, dtmax):
+    def __init__(self, universe, selection, t0=0, tf=-1, tau_max=20, verbose=False):
+
+        if tf >= len(universe.trajectory):
+            raise ValueError("\"end\" must be smaller than the number of frames in the trajectory.")
+
         self.universe = universe
         self.selection = selection
         self.t0 = t0
-        self.tf = tf
-        self.dtmax = dtmax
-        self.timeseries = []
-
-
-    def run(self):
-        """Analyze trajectory and produce timeseries"""
-
-        # select all frames to an array
-        selected = self._selection_serial(self.universe, self.selection)
-
-        if len(selected) < self.dtmax:
-            print ("ERROR: Cannot select fewer frames than dtmax")
-            return
-
-        for window_size in list(range(1, self.dtmax + 1)):
-            output = self._getMeanOnePoint(selected, window_size)
-            self.timeseries.append(output)
-
+        self.tf = len(list(universe.trajectory[:tf])) + 1
+        self.tau_max = tau_max
+        self.verbose = verbose
 
-    def _selection_serial(self, universe, selection_str):
-        selected = []
-        pm = ProgressMeter(self.tf-self.t0, interval=10,
-                           verbose=True, offset=-self.t0)
-        for ts in universe.trajectory[self.t0:self.tf]:
-            selected.append(universe.select_atoms(selection_str))
-            pm.echo(ts.frame)
-        return selected
+        if self.tau_max > self.tf - self.t0:
+            raise ValueError("Too few frames selected for given tau_max.")
 
+    def print(self, *args):
+        if self.verbose:
+            print(args)
 
-    def _getMeanOnePoint(self, selected, window_size):
+    def run(self):
         """
-        This function gets one point of the plot P(t) vs t. It uses the
-        _getOneDeltaPoint() function to calculate the average.
+        Computes and returns the survival probability timeseries
         """
-        n = 0
-        sumDeltaP = 0.0
-        for frame_no in range(len(selected) - window_size):
-            delta = self._getOneDeltaPoint(selected, frame_no, window_size)
-            sumDeltaP += delta
-            n += 1
+        
+        # load all frames to an array of sets
+        selected = []
+        for _ in self.universe.trajectory[self.t0:self.tf]:
+            selected.append(set(self.universe.select_atoms(self.selection).ids))
 
-        return sumDeltaP/n
+        tau_timeseries = list(range(1, self.tau_max + 1))
+        sp_timeseries = [[] for _ in range(self.tau_max)]
```

Comment:
So we do not know how many datapoints we are going to have ahead. This is because some reference frames will not contain the original molecule - hence they would be omitted.